src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+                {
+                    throw new ApplicationFailureException(
+                        "Failed decoding parameters", e);
+                }
+                // Append default parameters if needed
+                for (var i = tsk.Start.Input.Count; i < paramInfos.Length; i++)
+                {
+                    paramVals.Add(paramInfos[i].DefaultValue);
+                }
+
+                // Build the interceptor impls, chaining each interceptor in reverse
+                var inbound = worker.Interceptors.Reverse().Aggregate(
+                    (ActivityInboundInterceptor)new InboundImpl(),
+                    (v, impl) => impl.InterceptActivity(v));
+                // Initialize with outbound
+                inbound.Init(new OutboundImpl(this));
+
+                // Execute and put result on completed
+                var result = await inbound.ExecuteActivityAsync(new(
+                    Delegate: defn.Delegate,
+                    Parameters: paramVals.ToArray(),
+                    Headers: tsk.Start.HeaderFields)).ConfigureAwait(false);
+
+                completion.Result.Completed = new();
+                // As a special case, ValueTuple is considered "void"
+                if (result is not ValueTuple)
+                {
+                    completion.Result.Completed.Result =
+                        await worker.Client.Options.DataConverter.ToPayloadAsync(
+                            result).ConfigureAwait(false);
+                }
+            }
+            catch (CompleteAsyncException)
+            {
+                act.Context.Logger.LogDebug(
+                    "Completing activity {ActivityType} asynchronously",
+                    act.Context.Info.ActivityType);
+                completion.Result.WillCompleteAsync = new();
+            }
+            catch (OperationCanceledException) when (act.ServerRequestedCancel)
+            {
+                act.Context.Logger.LogDebug(
+                    "Completing activity {ActivityType} as cancelled",
+                    act.Context.Info.ActivityType);
+                completion.Result.Cancelled = new()
+                {
+                    Failure = await worker.Client.Options.DataConverter.ToFailureAsync(
+                        new CancelledFailureException("Cancelled")).ConfigureAwait(false),
+                };
+            }
+            catch (OperationCanceledException) when (act.HeartbeatFailureException != null)
+            {
+                act.Context.Logger.LogWarning(
+                    "Completing activity {ActivityType} as failure during heartbeat",
+                    act.Context.Info.ActivityType);
+                completion.Result.Failed = new()
+                {
+                    Failure_ = await worker.Client.Options.DataConverter.ToFailureAsync(
+                        act.HeartbeatFailureException).ConfigureAwait(false),
+                };
+            }
+            catch (Exception e)
```

Comment:
Should we have a case for when activity throws due to worker shutting down? At least for a nicer log message and possibly info log level.

Reply:
I do say how many activities I am force cancelling on worker shutdown in that log message, but I don't think we need a separate catch clause here for "cancelled failure + cancel reason is worker shutdown". These catch clauses are not for specific error messages (though they have them), they are for different behavior.

Reply:
I think it's still worth changing the logging in this case to reduce the chance of user confusion.

Reply:
Not following. It says why it failed in the log. The cancellation token was cancelled.

Reply:
Yeah, warning is fine here.